transforms/ExplicitConsdes.cpp

```diff
@@ -0,0 +1,231 @@
+                      "var initializer\n";
+            return;
+        }
+
+        for (unsigned i = 0; i < init->getType()->getArrayNumElements(); ++i) {
+            // type { i32, void ()*, i8* }
+            auto *elem = init->getAggregateElement(i);
+
+            FunctionEntry entry{};
+            entry.priority =
+                dyn_cast<ConstantInt>(elem->getAggregateElement(0u))
+                    ->getZExtValue();
+            entry.function = elem->getAggregateElement(1);
+            entry.param = elem->getAggregateElement(2);
+            target.push_back(std::move(entry));
+        }
+    }
+
+    void markUnused(GlobalVariable *gv) {
+        if (!gv)
+            return;
+
+        gv->setName(gv->getName() + "_unused");
+    }
+
+    void insertCalls(std::vector<FunctionEntry> &entries, Instruction *before) {
+        Instruction *after = nullptr;
+
+        for (auto &entry : entries) {
+            auto *call = CallInst::Create(
+                dyn_cast<FunctionType>(
+                    entry.function->getType()->getPointerElementType()),
+                entry.function, "");
+
+            call->setDebugLoc(before->getDebugLoc());
+
+            if (after) {
+                call->insertAfter(after);
+            } else {
+                call->insertBefore(before);
+            }
+            after = call;
+        }
+    }
+
+    bool isExit(const Function *f) {
+        const Type *ty = f->getType();
+        if (ty->isPointerTy())
+            ty = ty->getPointerElementType();
+
+        return f->hasName() && f->getName() == "exit" &&
+                    ty->getFunctionNumParams() == 1 &&
+                    ty->getFunctionParamType(0)->isIntegerTy();
+    }
+
+    void warnIfExitIsUsed(const Instruction& inst) {
+        // inst -> [operands] -> [only function pointers] -> [only exit] -> is_empty
+        for (const auto& op : inst.operands()) {
+            const auto *val = op.get();
+            if (!val)
+                continue;
+
+            const auto *ty = val->getType();
+
+            if (!ty->isPointerTy())
+                continue;
+
+            if (!ty->getPointerElementType()->isFunctionTy())
+                continue;
+
+            const auto *f = dyn_cast<Function>(val);
+            if (!f || !isExit(f))
+                continue;
+
+            errs() << "explicit-consdes: warning: indirect call of exit is possible in the program\n";
+        }
+    }
+
+   public:
+    static char ID;
+
+    ExplicitConsdes() : ModulePass(ID) {}
+
+    bool runOnModule(Module &mod) override {
+        using std::sort;
+        using std::tuple;
+        using std::vector;
+
+        bool change = false;
+
+        vector<FunctionEntry> ctors, dtors;
+
+        GlobalVariable *dtorsVar = nullptr;
+        GlobalVariable *ctorsVar = nullptr;
+
+        for (auto &glob : mod.globals()) {
+            if (glob.hasName() && glob.getName() == "llvm.global_dtors") {
```

Comment:
Oh, I overlooked this. llvm::Module has `getNamedGlobal(StringRef)` method, you can use that to obtain the globals without iterating over them.